src/storage/mod.rs

```diff
@@ -0,0 +1,262 @@
+        }
+    }
+
+    Ok(files)
+}
+
+pub(crate) enum Storage<'a> {
+    Database(DatabaseBackend<'a>),
+    S3(S3Backend<'a>),
+}
+
+impl<'a> Storage<'a> {
+    pub(crate) fn new(conn: &'a Connection) -> Self {
+        if let Some(c) = s3::s3_client() {
+            Storage::from(S3Backend::new(c, s3::S3_BUCKET_NAME))
+        } else {
+            DatabaseBackend::new(conn).into()
+        }
+    }
+    pub(crate) fn get(&self, path: &str) -> Result<Blob, Error> {
+        match self {
+            Self::Database(db) => db.get(path),
+            Self::S3(s3) => s3.get(path),
+        }
+    }
+
+    fn store_batch(&mut self, batch: &[Blob], trans: &Transaction) -> Result<(), Error> {
+        match self {
+            Self::Database(db) => db.store_batch(batch, trans),
+            Self::S3(s3) => s3.store_batch(batch),
+        }
+    }
+
+    // Store all files in `root_dir` into the backend under `prefix`.
+    //
+    // If the environment is configured with S3 credentials, this will upload to S3;
+    // otherwise, this will store files in the database.
+    //
+    // This returns a HashMap<filename, mime type>.
+    pub(crate) fn store_all(
+        &mut self,
+        conn: &Connection,
+        prefix: &str,
+        root_dir: &Path,
+    ) -> Result<HashMap<PathBuf, String>, Error> {
+        const MAX_CONCURRENT_UPLOADS: usize = 1000;
+
+        let trans = conn.transaction()?;
+        let mut file_paths_and_mimes = HashMap::new();
+
+        get_file_list(root_dir)?
+            .into_iter()
+            .filter_map(|file_path| {
+                // Some files have insufficient permissions
+                // (like .lock file created by cargo in documentation directory).
+                // Skip these files.
+                fs::File::open(root_dir.join(&file_path))
+                    .ok()
+                    .map(|file| (file_path, file))
+            })
+            .map(|(file_path, mut file)| {
+                let mut content: Vec<u8> = Vec::new();
+                file.read_to_end(&mut content)?;
+
+                let bucket_path = Path::new(prefix).join(&file_path);
+
+                #[cfg(windows)] // On windows, we need to normalize \\ to / so the route logic works
+                let bucket_path = path_slash::PathBufExt::to_slash(&bucket_path).unwrap();
+                #[cfg(not(windows))]
+                let bucket_path = bucket_path.into_os_string().into_string().unwrap();
+
+                let mime = detect_mime(&file_path)?;
+
+                file_paths_and_mimes.insert(file_path, mime.to_string());
+                Ok(Blob {
+                    path: bucket_path,
+                    mime: mime.to_string(),
+                    content,
+                    // this field is ignored by the backend
+                    date_updated: Timespec::new(0, 0),
+                })
+            })
+            .collect::<Result<Vec<_>, Error>>()?
```

Comment:
Hmm, I just realized this will load all the files into memory at once. Do we want me to try and make this lazy so it instead only loads a batch at a time?